README.md

# base-repository

[![npm version](https://img.shields.io/npm/v/easy-template.svg)](https://www.npmjs.com/package/easy-template)
[![Build Status](https://img.shields.io/travis/LiuYueKai/easy-template/master.svg)](https://travis-ci.org/LiuYueKai/easy-template)
[![devDependency Status](https://img.shields.io/david/dev/LiuYueKai/easy-template.svg)](https://david-dm.org/LiuYueKai/easy-template#info=devDependencies)
[![NPM downloads](http://img.shields.io/npm/dm/easy-template.svg?style=flat)](https://npmjs.org/package/easy-template)

[中文文档](https://github.com/LiuYueKai/easy-template/blob/master/README_CN.md)

## Introduction
This is a template creation tool for web engineering, through ` et init ` selected template quickly generate engineering.


## Quickstart

### Installation
```
npm install easy-template -g
```

### Usage

```sh
$ et init
```
1. Open terminal,input `et init` view.
2. Displays the boilerplate names and descriptions available online.
3. Keyboard select what you need  boilerplate.( `↑` , `↓` )

## Contribution
Please make sure to read the Contributing Guide before making a pull request.

## Feedback

If you encounter any problems , submit [issues](https://github.com/LiuYueKai/easy-template/issues)。

## Licence
[MIT](https://github.com/LiuYueKai/easy-template/blob/master/LICENSE)


CHANGELOG.md
<a name="1.0.2"></a>
## [1.0.2](https://github.com/LiuYueKai/easy-template/compare/v1.0.1...v1.0.2) (2017-04-17)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/LiuYueKai/easy-template/compare/v1.0.0...v1.0.1) (2017-04-17)



<a name="1.0.0"></a>
# 1.0.0 (2017-04-17)





lib/help.js
/**
 * Module : helper
 * Author : LiuYueKai(liuyuekainihao@163.com)
 * Date	  : 2017-04-17 14:38:42
 */

'use strict';

const chalk = require('chalk');
const path = require('path');
const fs = require('fs');

module.exports = {
    help: () => {
        console.log(chalk.green('et help'));
        console.log(chalk.green('1. et init                  Generate best practices'));
        console.log(chalk.green('2. et -h	             Help'));
    },
    version: () => {
        console.log();
        console.log(chalk.green('Version : ' + require('../package.json').version));
        console.log();
        process.exit();
    },
    info: (msg) => {
        console.log(chalk.cyan("Info : " + msg));
    },
    error: (msg) => {
        console.log(chalk.red("Error : " + msg));
    },
    getInitGithubRepo: (repo) => {
        var _ut = 'uba-templates/',
            _repo;
        if (repo.indexOf('/') !== -1) {
            _repo = repo;
        } else {
            _repo = _ut + repo;
        }
        return _repo;
    }
}


lib/init.js
/**
 * Module : et init 
 * Author : LiuYueKai(liuyuekainihao@163.com)
 * Date	  : 2017-04-17 17:06:42
 */

const fs = require('fs');
const request = require('request');
const chalk = require('chalk');
const inquirer = require('inquirer');
const path = require('path');
const pathExists = require('path-exists');
const download = require('download-git-repo');
const spawn = require('cross-spawn');
const help = require('./help');

module.exports = () => {
    help.info("Available templates:");
    var repoNameData = [];
    request({
        url: 'https://api.github.com/users/easy-templates/repos',
        headers: {
            'User-Agent': 'easy templates'
        }
    }, function(err, res, body) {
        if (err) console.log(err);
        var requestBody = JSON.parse(body);
        if (Array.isArray(requestBody)) {
            requestBody.forEach(function(repo, index) {
                repoNameData.push(`${repo.name} - ${repo.description}`);
            });
            //TODO 人机交互
            inquirer.prompt([{
                type: 'list',
                name: 'selectRepo',
                message: 'Please select :',
                choices: repoNameData
            }]).then(function(answers) {
                var selectName = answers.selectRepo.split(' - ')[0];
                var questions = [{
                    type: 'input',
                    name: 'selectName',
                    message: 'boilerplate name :',
                    default: function() {
                        return 'base-repository';
                    }
                }];
                inquirer.prompt(questions).then(function(answers) {
                    var name = answers.selectName,
                        template = selectName;
                    var root = path.resolve(name);
                    if (!pathExists.sync(name)) {
                        fs.mkdirSync(root);
                    } else {
                        help.error(`directory ${name} already exists.`);
                        process.exit(0);
                    }
                    help.info(`Downloading ${template} please wait.`);
                    //TODO 开始下载
                    download('easy-templates/' + template, `${name}`, function(err) {
                        if (!err) {
                            help.info(`Boilerplate ${name} done.`);
                            inquirer.prompt([{
                                type: 'confirm',
                                message: 'Automatically install NPM dependent packages?',
                                name: 'ok'
                            }]).then(function(res) {
                                var npmInstallChdir = path.resolve('.', name);
                                if (res.ok) {
                                    help.info(`Install NPM dependent packages,please wait.`);
                                    //TODO 选择自动安装
                                    process.chdir(npmInstallChdir);
                                    var args = ['install'].filter(function(e) {
                                        return e;
                                    });
                                    var proc = spawn('npm', args, {
                                        stdio: 'inherit'
                                    });
                                    proc.on('close', function(code) {
                                        if (code !== 0) {
                                            console.error('`npm ' + args.join(' ') + '` failed');
                                            return;
                                        }
                                        help.info(`NPM package installed.`);
                                    });

                                } else {
                                    help.info(`Cancel the installation of NPM dependent package.\nPlease run \'cd ${name} && npm install\' manually.`);
                                }

                            });
                        } else {
                            console.error(requestBody.message);
                        }
                    });
                });
            });



        }
    });
}
